### PR TITLE
🛠️  [maintenance] Template Migration

### PR DESCRIPTION
**Pull request by JacobMGEvans**
Moving template from its own repo into the Wrangler monorepo for more direct support, maintenance, and improvements. It will also allow for additional Wrangler integration with 'init' command and E2E testing.

Related Docs PR: https://github.com/cloudflare/cloudflare-docs/pull/7242

Additional Tasks for completing this PR:
- [x] The templates themselves will moved into the repo root.
- [x] The packages within `/templates` will be moved into the `wrangler2/packages` directly. 
- [ ] Testing and custom CI will be made for the two packages 
- [ ] The custom CI is for `worker.new` which is a Worker using Wrangler, and was previously manually deployed.

After this PR: 
- [ ]  Migrating and archiving templates that are their own repos still. 
- [ ]  Publishing the updated Worker that runs worker.new

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,13 +11,14 @@
...
-		"fixtures/*"
+		"fixtures/*",
+		"packages/templates/packages/create-cloudflare"
 	],
 	"scripts": {
 		"build": "npm run build --workspace=wrangler --workspace=jest-environment-wrangler --workspace=pages-plugin-example --workspace=wranglerjs-compat-webpack-plugin",
 		"check": "run-p check:* --aggregate-output --continue-on-error",
 		"check:format": "prettier . --check --ignore-unknown",
-		"check:lint": "eslint \"packages/**/*.[tj]s?(x)\" --cache --cache-strategy content --max-warnings=0",
+		"check:lint": "eslint \"packages/**/*.[tj]s?(x)\" --cache --cache-strategy content --max-warnings=0 --config ./.eslintrc.js",
```
Comment (JacobMGEvans) [inline package.json:21]: ESlint CLI was not using the config properly, explicitly defining the config to use in ESLint CLI [2c28cd2](https://github.com/cloudflare/wrangler2/pull/2498/commits/2c28cd24fa4b1e7aba267052c204f308ea94029f)

File: packages/templates/packages/workers.new/src/index.ts [hunk 2]
```diff
@@ -5,11 +5,11 @@
 type Redirects = Record<string, [string, string, string, string?]>;
 
 // stackblitz repository source
-const source = "github/cloudflare/templates/tree/main";
+const source = "github/wrangler2/tree/main/packages/templates";
 
 // deploy with cloudflare source
 
-const src = "https://github.com/cloudflare/templates/tree/main";
```
Comment (JacobMGEvans) [inline packages/templates/packages/workers.new/src/index.ts]: Need to do a `wrangler publish` on this Worker at some point after the PR goes in.

File: package.json [hunk 3]
```diff
@@ -12,7 +12,7 @@
 	"workspaces": [
 		"packages/*",
 		"fixtures/*",
-		"packages/templates"
+		"packages/templates/packages/create-cloudflare"
```
Comment (JacobMGEvans) [inline package.json:15]: Due to `changesets` using workspaces to determine what to check for changes, we have to specify the only deployable package `create-cloudflare` otherwise changesets breaks on `"packages/templates"` reason being all the templates are missing `"name"` & `"version"` in all the `package.json`

File: packages/templates/pages-example-forum-app/tsconfig.json [hunk 4]
```diff
@@ -98,7 +98,7 @@
 
 		/* Completeness */
 		// "skipDefaultLibCheck": true,                      /* Skip type checking .d.ts files that are included with TypeScript. */
-		"skipLibCheck": true /* Skip type checking all .d.ts files. */,
-		"exclude": ["functions/**/*.{js, ts}"]
-	}
+		"skipLibCheck": true /* Skip type checking all .d.ts files. */
+	},
+	"exclude": ["functions/**/*.{js, ts}"]
```
Comment (JacobMGEvans) [inline packages/templates/pages-example-forum-app/tsconfig.json:103]: `exclude` was in the compilerOption breaking the TS/Lint check

File: packages/templates/packages/workers.new/src/index.ts [hunk 5]
```diff
@@ -5,11 +5,11 @@
...
+const src = "https://github.com/wrangler2/tree/main/packages/templates";
```
Comment (Skye-31) [inline packages/templates/packages/workers.new/src/index.ts:12]: Should this not be `github.com/cloudflare/wrangler2`?

File: packages/templates/packages/workers.new/src/index.ts [hunk 6]
```diff
@@ -5,11 +5,11 @@
 type Redirects = Record<string, [string, string, string, string?]>;
 
 // stackblitz repository source
-const source = "github/cloudflare/templates/tree/main";
+const source = "github/wrangler2/tree/main/packages/templates";
```
Comment (Skye-31) [inline packages/templates/packages/workers.new/src/index.ts:8]: This should also include `/cloudflare/`

Comment (JacobMGEvans) [reply packages/templates/packages/workers.new/src/index.ts:8]: lol oops, nice catch

**Comment by Skye-31** [comment]
I can't actually review - not a maintainer, but LGTM

File: package.json (hunk 3 above)
Comment (petebacondarwin) [reply package.json:15]: I think what we actually want is to move the packages/templates/packages/create-cloudflare to packages/create-cloudflare, right?

Comment (petebacondarwin) [reply package.json:15]: Along with `workers.new`

Comment (JacobMGEvans) [reply package.json:15]: 🤔

**Comment by JacobMGEvans** [comment]
https://github.com/cloudflare/wrangler2/pull/2498#discussion_r1069955678
> I think what we actually want is to move the `packages/templates/packages/create-cloudflare` to `packages/create-cloudflare`, right?


Based on internal conversation: 
1. The templates themselves will moved into the repo root.
2. The packages within `/templates` will be moved into the `wrangler2/packages` directly. 
3. Testing and custom CI will be made for the two packages 
    a. The custom CI is for `worker.new` which is a Worker using Wrangler, and was previously manually deployed.

@petebacondarwin Feel free to add anything I may have missed from our conversation.

**Comment by petebacondarwin** [comment]
> The templates themselves will moved into the repo root.\

To be 100% clear. The templates would end up in `/templates` in this repo (not directly in the root, right?

**Comment by JacobMGEvans** [comment]
> > The templates themselves will moved into the repo root.\
> 
> To be 100% clear. The templates would end up in `/templates` in this repo (not directly in the root, right?

Yes. That is was I mean 😆

File: package.json (hunk 3 above)
Comment (penalosa) [reply package.json:15]: Looks like this needs changed after the move of templates to the root?

File: tsconfig.json [hunk 7]
```diff
@@ -14,8 +14,7 @@
 		// `alwaysStrict: false` is needed because otherwise esbuild will add `"use strict";`
 		// statements throughout the imported libraries which can break them.
 		"alwaysStrict": false,
-		"skipLibCheck": true,
-		"types": []
```
Comment (penalosa) [inline tsconfig.json:18]: Why this change?

**Review by penalosa — CHANGES_REQUESTED** [review]
Couple changes, primarily to paths

File: packages/create-cloudflare/.gitignore [hunk 8]
```diff
@@ -0,0 +1,8 @@
+.DS_Store
+node_modules
```
Comment (petebacondarwin) [inline packages/create-cloudflare/.gitignore:2]: Perhaps as a cleanup later on, but these are already covered in the top level .gitignore, and generally we have been trying to avoid adding further .gitignores to the monorepo. For example the top-level one contains specific ignores for packages/wrangler, packages/prerelease-registry and packages/wranglerjs-compat-webpack-plugin - although I note that wrangler-devtools sneaked its own file in recently...

File: packages/create-cloudflare/readme.md [hunk 9]
```diff
@@ -0,0 +1,70 @@
+# create-cloudflare
+
+> Create new Cloudflare projects with one command
+
+> **Note**
+> This package has now been sunsetted! We recommend using the `wrangler generate [name] [template]` command to create new projects with templates.
```
Comment (petebacondarwin) [inline packages/create-cloudflare/readme.md:6]: If we have sunsetted this package, perhaps we should not bring it in to this monorepo?

File: packages/create-cloudflare/src/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,7 @@
+export interface Argv {
+	init?: boolean;
+	force?: boolean;
+	debug?: boolean;
+}
```
Comment (petebacondarwin) [inline packages/create-cloudflare/src/index.d.ts:5]: Feels like this file ought to be generated at build/publish time and not checked in?

File: packages/create-cloudflare/tsconfig.json [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+		"strictNullChecks": true,
+		"noImplicitThis": true,
+		"noImplicitAny": true,
+		"alwaysStrict": true,
+		"module": "esnext",
+		"allowJs": false,
+		"noEmit": true,
+		"types": ["node"],
+		"paths": {
+			"create-cloudflare": ["./src/index.d.ts"]
```
Comment (petebacondarwin) [inline packages/create-cloudflare/tsconfig.json:22]: This seems rather circular?

File: packages/create-cloudflare/src/index.d.ts (hunk 10 above)
Comment (petebacondarwin) [reply packages/create-cloudflare/src/index.d.ts:5]: And probably move the `Argv` type to index.ts or utils.ts?

File: templates/.gitignore [hunk 12]
```diff
@@ -0,0 +1,5 @@
+.DS_Store
+node_modules
+*-lock.*
+*.lock
+*.log
```
Comment (petebacondarwin) [inline templates/.gitignore:5]: Can we just move these to the top level .gitignore, if they are not there already? (Ensuring we don't ignore our top level package-lock.json!)

File: templates/.prettierignore [hunk 13]
```diff
@@ -0,0 +1,10 @@
+*-lock.*
+*.lock
+*.log
+
+/tsconfig.json
+
+build/
+dist/
+
+/worker-rust/target/
```
Comment (petebacondarwin) [inline templates/.prettierignore:10]: Can we move these to the top-level .prettierignore?

File: templates/tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+		"resolveJsonModule": true,
+		"strictNullChecks": true,
+		"noImplicitThis": true,
+		"noImplicitAny": true,
+		"alwaysStrict": true,
+		"module": "esnext",
+		"checkJs": true,
+		"noEmit": true,
+		"paths": {
+			"create-cloudflare": ["./packages/create-cloudflare/src/index.d.ts"]
```
Comment (petebacondarwin) [inline templates/tsconfig.json:22]: Hmm....

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Looking good. I haven't yet pulled it down and tried testing this stuff...

**Comment by petebacondarwin** [comment]
```sh
npm run build -w create-cloudflare

> create-cloudflare@1.0.0 build
> bundt -m

sh: bundt: command not found
npm ERR! Lifecycle script `build` failed with error: 
npm ERR! Error: command failed 
npm ERR!   in workspace: create-cloudflare@1.0.0 
npm ERR!   at location: /Users/pbacondarwin/dev/wrangler2/packages/create-cloudflare 
```

File: packages/create-cloudflare/readme.md [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+$ pnpm create cloudflare <directory> <source> [options]
+# or
+$ yarn create cloudflare <directory> <source> [options]
+```
+
+> **Note:** All recent versions of `npm`, `yarn`, and `pnpm` support this feature!
+
+### Sources
+
+You may select the name of any subdirectory within the [`cloudflare/templates`](https://github.com/cloudflare/templates) repository to create your project; for example, `worker-typescript` and `examples/fast-google-fonts` are both valid subdirectory names.
```
Comment (petebacondarwin) [inline packages/create-cloudflare/readme.md:24]: The URL to cloudflare/templates needs updating

**Comment by petebacondarwin** [comment]
Looks like the build for `create-cloudflare` is totally broken.
Either we need to fix that, or remove this package from the repo, IMO.

File: package.json [hunk 16]
```diff
@@ -11,13 +11,14 @@
 	"author": "wrangler@cloudflare.com",
 	"workspaces": [
 		"packages/*",
-		"fixtures/*"
+		"fixtures/*",
+		"packages/templates/packages/create-cloudflare"
```
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply package.json:15]: Oh nice catch I missed that, thanks.

File: packages/create-cloudflare/.gitignore (hunk 8 above)
Comment (JacobMGEvans) [reply packages/create-cloudflare/.gitignore:2]: Will remove then.

**Comment by petebacondarwin** [comment]
We should add `packages/workers.new` to the vistest.config.js includes.

File: packages/create-cloudflare/readme.md (hunk 9 above)
Comment (JacobMGEvans) [reply packages/create-cloudflare/readme.md:6]: If we are no longer implementing it in generate, just using our own code than I agree with this and we can avoid the unnecessary maintenance burden.

File: templates/.prettierignore (hunk 13 above)
Comment (JacobMGEvans) [reply templates/.prettierignore:10]: I wasn't sure if they should be scoped or top-level I definitely not against moving them up.

File: tsconfig.json (hunk 7 above)
Comment (JacobMGEvans) [reply tsconfig.json:18]: empty `types` doesnt mean anything

as for `skipLibCheck` I rather not skip type checking like this if we don't have to.

File: .eslintrc.js [hunk 17]
```diff
@@ -112,6 +115,7 @@ module.exports = {
 		"packages/jest-environment-wrangler/dist",
 		"packages/wrangler-devtools/built-devtools",
 		"packages/wranglerjs-compat-webpack-plugin/lib",
+		"packages/templates",
```
Comment (penalosa) [inline .eslintrc.js:118]: Should this be `/templates`?

File: templates/examples/cryptocurrency-slack-bot/index.js [hunk 18]
```diff
@@ -0,0 +1,302 @@
+addEventListener("fetch", (event) => {
+	event.respondWith(slackWebhookHandler(event.request));
+});
+
+// SLACK_TOKEN is used to authenticate requests are from Slack.
+// Keep this value secret.
+let SLACK_TOKEN = "PUTYOURTOKENHERE";
```
Comment (penalosa) [inline templates/examples/cryptocurrency-slack-bot/index.js:7]: Uhhhhh. This should be a secret binding. Not a blocker for this PR, obviously (this being copied over from the templates repo), but these could probably do with a security review at some point

File: tsconfig.json (hunk 7 above)
Comment (penalosa) [reply tsconfig.json:18]: An empty `types` means that no `@types` packages will be automatically included, I think? Whereas _no_ `types` will include them automatically, which we probably don't want to do (I believe this was causing issues in the past with `@types/node` being used in the wrong places cc @mrbbot )

**Review by penalosa — COMMENTED** [review]
Terribly sorry, it seems my review comments from yesterday weren't sent 😂

File: package.json (hunk 16 above)
[the remark this answers is not in the document]
Comment (JacobMGEvans) [reply package.json:15]: I added `templates` since it is similar to fixtures

File: packages/create-cloudflare/readme.md (hunk 9 above)
Comment (JacobMGEvans) [reply packages/create-cloudflare/readme.md:6]: Removed the package!

File: packages/create-cloudflare/tsconfig.json (hunk 11 above)
Comment (JacobMGEvans) [reply packages/create-cloudflare/tsconfig.json:22]: Removed with the package. 😄

File: packages/create-cloudflare/src/index.d.ts (hunk 10 above)
Comment (JacobMGEvans) [reply packages/create-cloudflare/src/index.d.ts:5]: package removed.

File: templates/.gitignore (hunk 12 above)
Comment (JacobMGEvans) [reply templates/.gitignore:5]: Moved the lock related ones and should only affect the locks in templates.

File: packages/create-cloudflare/readme.md (hunk 15 above)
Comment (JacobMGEvans) [reply packages/create-cloudflare/readme.md:24]: Package removed.

File: .eslintrc.js (hunk 17 above)
Comment (JacobMGEvans) [reply .eslintrc.js:118]: Thanks!

File: templates/examples/cryptocurrency-slack-bot/index.js (hunk 18 above)
Comment (JacobMGEvans) [reply templates/examples/cryptocurrency-slack-bot/index.js:7]: Should make this a ticket.

File: templates/tsconfig.json (hunk 14 above)
Comment (JacobMGEvans) [reply templates/tsconfig.json]: removed

**Review by penalosa — APPROVED** [review]

**Review by penalosa — CHANGES_REQUESTED** [review]
Only remaining things:
- https://github.com/cloudflare/workers-sdk/pull/2498#issuecomment-1405682082
- Looks like two new templates have been added: `worker-prospector` and `worker-r2`

**Review by penalosa — APPROVED** [review]
